principal/server.go

```diff
@@ -131,56 +137,70 @@ func NewServer(ctx context.Context, kubeClient *kube.KubernetesClient, namespace
+		}),
+		informer.WithWatchHandler[*v1alpha1.Application](func(ctx context.Context, opts v1.ListOptions) (watch.Interface, error) {
+			return kubeClient.ApplicationsClientset.ArgoprojV1alpha1().Applications("").Watch(ctx, opts)
+		}),
+		informer.WithAddHandler[*v1alpha1.Application](s.newAppCallback),
+		informer.WithUpdateHandler[*v1alpha1.Application](s.updateAppCallback),
+		informer.WithDeleteHandler[*v1alpha1.Application](s.deleteAppCallback),
+		informer.WithFilters[*v1alpha1.Application](appFilters),
 	}
 
-	managerOpts := []application.ApplicationManagerOption{
+	appManagerOpts := []application.ApplicationManagerOption{
 		application.WithAllowUpsert(true),
 		application.WithRole(manager.ManagerRolePrincipal),
 	}
 
-	appProjectInformerOptions := []appprojectinformer.AppProjectInformerOption{
-		appprojectinformer.WithAddFunc(s.newAppProjectCallback),
-		appprojectinformer.WithUpdateFunc(s.updateAppProjectCallback),
-		appprojectinformer.WithDeleteFunc(s.deleteAppProjectCallback),
+	projInformerOpts := []informer.InformerOption[*v1alpha1.AppProject]{
+		informer.WithListHandler[*v1alpha1.AppProject](func(ctx context.Context, opts v1.ListOptions) (runtime.Object, error) {
+			return kubeClient.ApplicationsClientset.ArgoprojV1alpha1().AppProjects("").List(ctx, opts)
+		}),
+		informer.WithWatchHandler[*v1alpha1.AppProject](func(ctx context.Context, opts v1.ListOptions) (watch.Interface, error) {
+			return kubeClient.ApplicationsClientset.ArgoprojV1alpha1().AppProjects("").Watch(ctx, opts)
+		}),
+		informer.WithAddHandler[*v1alpha1.AppProject](s.newAppProjectCallback),
+		informer.WithUpdateHandler[*v1alpha1.AppProject](s.updateAppProjectCallback),
+		informer.WithDeleteHandler[*v1alpha1.AppProject](s.deleteAppProjectCallback),
 	}
 
-	appProjectManagerOption := []appproject.AppProjectManagerOption{
+	projManagerOpts := []appproject.AppProjectManagerOption{
 		appproject.WithAllowUpsert(true),
 		appproject.WithRole(manager.ManagerRolePrincipal),
 	}
 
 	if s.options.metricsPort > 0 {
-		appInformerOptions = append(appInformerOptions, appinformer.WithMetrics(metrics.NewApplicationWatcherMetrics()))
-		managerOpts = append(managerOpts, application.WithMetrics(metrics.NewApplicationClientMetrics()))
-		appProjectInformerOptions = append(appProjectInformerOptions, appprojectinformer.WithMetrics(metrics.NewAppProjectWatcherMetrics()))
-		appProjectManagerOption = append(appProjectManagerOption, appproject.WithMetrics(metrics.NewAppProjectClientMetrics()))
+		appInformerOpts = append(appInformerOpts, informer.WithMetrics[*v1alpha1.Application](prometheus.NewRegistry(), metrics.NewInformerMetrics("applications")))
+		appManagerOpts = append(appManagerOpts, application.WithMetrics(metrics.NewApplicationClientMetrics()))
+		projInformerOpts = append(projInformerOpts, informer.WithMetrics[*v1alpha1.AppProject](prometheus.NewRegistry(), metrics.NewInformerMetrics("applications")))
```

Comment:
Should the label be `appProjects` instead of `applications`?

Reply:
Good catch!